File: samples/java/spark/sources/SparkQR.java
/* file: SparkQR.java */
/*******************************************************************************
* Copyright 2017-2019 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
 //  Content:
 //     Java sample of computing QR decomposition in the distributed processing
 //     mode
 ////////////////////////////////////////////////////////////////////////////////
 */

package DAAL;

import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;

import org.apache.spark.api.java.*;
import org.apache.spark.api.java.function.*;
import org.apache.spark.SparkConf;

import scala.Tuple2;

import com.intel.daal.data_management.data.HomogenNumericTable;
import com.intel.daal.data_management.data.NumericTable;
import com.intel.daal.algorithms.qr.*;
import com.intel.daal.data_management.data.*;
import com.intel.daal.data_management.data_source.FileDataSource;
import com.intel.daal.data_management.data_source.DataSource;
import com.intel.daal.services.*;

public class SparkQR {
    /* Class containing the algorithm results */
    static class QRResult {
        public HomogenNumericTable R;
        public JavaPairRDD<Integer, HomogenNumericTable> Q;
    }

    static QRResult result = new QRResult();

    static JavaPairRDD<Integer, DataCollection> dataFromStep1ForStep2_RDD;
    static JavaPairRDD<Integer, DataCollection> dataFromStep1ForStep3_RDD;
    static JavaPairRDD<Integer, DataCollection> dataFromStep2ForStep3_RDD;

    public static QRResult runQR(DaalContext context, JavaPairRDD<Integer, HomogenNumericTable> dataRDD, JavaSparkContext sc) {

        computeStep1Local(dataRDD);

        computeStep2Master(context, sc);

        computeStep3Local();

        return result;
    }

    private static void computeStep1Local(JavaPairRDD<Integer, HomogenNumericTable> dataRDD) {
        /* Create an RDD containing partial results for steps 2 and 3 */
        JavaPairRDD<Integer, Tuple2<DataCollection, DataCollection>> dataFromStep1_RDD = dataRDD.mapToPair(
        new PairFunction<Tuple2<Integer, HomogenNumericTable>, Integer, Tuple2<DataCollection, DataCollection>>() {
            public Tuple2<Integer, Tuple2<DataCollection, DataCollection>> call(Tuple2<Integer, HomogenNumericTable> tup) {
                DaalContext context = new DaalContext();

                /* Create an algorithm to compute QR decomposition on local nodes */
                DistributedStep1Local qrStep1Local = new DistributedStep1Local(context, Double.class, Method.defaultDense);
                tup._2.unpack(context);
                qrStep1Local.input.set( InputId.data, tup._2 );

                /* Compute QR decomposition in step 1  */
                DistributedStep1LocalPartialResult pres = qrStep1Local.compute();
                DataCollection dataFromStep1ForStep2 = pres.get( PartialResultId.outputOfStep1ForStep2 );
                dataFromStep1ForStep2.pack();
                DataCollection dataFromStep1ForStep3 = pres.get( PartialResultId.outputOfStep1ForStep3 );
                dataFromStep1ForStep3.pack();

                context.dispose();

                return new Tuple2<Integer, Tuple2<DataCollection, DataCollection>>(
                           tup._1, new Tuple2<DataCollection, DataCollection>(dataFromStep1ForStep2, dataFromStep1ForStep3));
            }
        }).cache();

        /* Extract partial results for step 3 */
        dataFromStep1ForStep3_RDD = dataFromStep1_RDD.mapToPair(
        new PairFunction<Tuple2<Integer, Tuple2<DataCollection, DataCollection>>, Integer, DataCollection>() {
            public Tuple2<Integer, DataCollection> call(Tuple2<Integer, Tuple2<DataCollection, DataCollection>> tup) {
                return new Tuple2<Integer, DataCollection>(tup._1, tup._2._2);
            }
        });

        /* Extract partial results for step 2 */
        dataFromStep1ForStep2_RDD = dataFromStep1_RDD.mapToPair(
        new PairFunction<Tuple2<Integer, Tuple2<DataCollection, DataCollection>>, Integer, DataCollection>() {
            public Tuple2<Integer, DataCollection> call(Tuple2<Integer, Tuple2<DataCollection, DataCollection>> tup) {
                return new Tuple2<Integer, DataCollection>(tup._1, tup._2._1);
            }
        });
    }

    private static void computeStep2Master(DaalContext context, JavaSparkContext sc) {

        int nBlocks = (int)(dataFromStep1ForStep2_RDD.count());

        List<Tuple2<Integer, DataCollection>> dataFromStep1ForStep2_List = dataFromStep1ForStep2_RDD.collect();

        /* Create an algorithm to compute QR decomposition on the master node */
        DistributedStep2Master qrStep2Master = new DistributedStep2Master( context, Double.class, Method.defaultDense );

        for (Tuple2<Integer, DataCollection> value : dataFromStep1ForStep2_List) {
            value._2.unpack(context);
            qrStep2Master.input.add( DistributedStep2MasterInputId.inputOfStep2FromStep1, value._1, value._2 );
        }

        /* Compute QR decomposition in step 2 */
        DistributedStep2MasterPartialResult pres = qrStep2Master.compute();

        KeyValueDataCollection inputForStep3FromStep2 = pres.get( DistributedPartialResultCollectionId.outputOfStep2ForStep3 );

        List<Tuple2<Integer, DataCollection>> list = new ArrayList<Tuple2<Integer, DataCollection>>(nBlocks);
        for (Tuple2<Integer, DataCollection> value : dataFromStep1ForStep2_List) {
            DataCollection dc = (DataCollection) inputForStep3FromStep2.get( value._1 );
            dc.pack();
            list.add(new Tuple2<Integer, DataCollection>(value._1, dc));
        }

        /* Make PairRDD from the list */
        dataFromStep2ForStep3_RDD = JavaPairRDD.fromJavaRDD( sc.parallelize(list, nBlocks) );

        Result res = qrStep2Master.finalizeCompute();

        HomogenNumericTable ntR = (HomogenNumericTable)res.get( ResultId.matrixR );
        result.R = ntR;
    }

    private static void computeStep3Local() {
        /* Group partial results from steps 1 and 2 */
        JavaPairRDD<Integer, Tuple2<Iterable<DataCollection>, Iterable<DataCollection>>> dataForStep3_RDD =
            dataFromStep1ForStep3_RDD.cogroup(dataFromStep2ForStep3_RDD);

        JavaPairRDD<Integer, HomogenNumericTable> ntQ_RDD = dataForStep3_RDD.mapToPair(
        new PairFunction<Tuple2<Integer, Tuple2<Iterable<DataCollection>, Iterable<DataCollection>>>, Integer, HomogenNumericTable>() {

            public Tuple2<Integer, HomogenNumericTable>
            call(Tuple2<Integer, Tuple2<Iterable<DataCollection>, Iterable<DataCollection>>> tup) {
                DaalContext context = new DaalContext();

                DataCollection ntQPi = tup._2._1.iterator().next();
                ntQPi.unpack(context);
                DataCollection ntPi  = tup._2._2.iterator().next();
                ntPi.unpack(context);

                /* Create an algorithm to compute QR decomposition on the master node */
                DistributedStep3Local qrStep3Local = new DistributedStep3Local(context, Double.class, Method.defaultDense);
                qrStep3Local.input.set( DistributedStep3LocalInputId.inputOfStep3FromStep1, ntQPi );
                qrStep3Local.input.set( DistributedStep3LocalInputId.inputOfStep3FromStep2, ntPi );

                /* Compute QR decomposition in step 3 */
                qrStep3Local.compute();
                Result result = qrStep3Local.finalizeCompute();

                HomogenNumericTable Qi = (HomogenNumericTable)result.get( ResultId.matrixQ );
                Qi.pack();

                context.dispose();

                return new Tuple2<Integer, HomogenNumericTable>(tup._1, Qi);
            }
        });
        result.Q = ntQ_RDD;
    }
}

File: samples/python/spark/sources/spark_QR.py
# file: spark_QR.py
#===============================================================================
# Copyright 2017-2019 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#===============================================================================

#
#  Content:
#     Python sample of computing QR decomposition in the distributed processing
#     mode
#
from __future__ import print_function
import os
import sys

from pyspark import SparkContext
from pyspark.conf import SparkConf

from daal import step1Local, step2Master, step3Local
from daal.algorithms import qr

from distributed_hdfs_dataset import (
    DistributedHDFSDataSet, deserializeNumericTable, serializeNumericTable, deserializeDataCollection
)

utils_folder = os.path.join(os.environ['DAALROOT'], 'examples', 'python', 'source')
if utils_folder not in sys.path:
    sys.path.insert(0, utils_folder)
from utils import printNumericTable


def runQR(dataRDD, sc):

    partsRDD = computeStep1Local(dataRDD)
    res = computeStep2Master(partsRDD['1for2'])
    q = computeStep3Local(partsRDD['1for3'], res['from2_for3'])

    return {'Q': q, 'R': res['ntR']}


def computeStep1Local(dataRDD):
    # Create an RDD containing partial results for steps 2 and 3
    def mapper(tup):

        key, homogen_table = tup

        # Create an algorithm to compute QR decomposition on local nodes
        qrStep1Local = qr.Distributed(step1Local, method=qr.defaultDense)
        deserialized_homogen_table = deserializeNumericTable(homogen_table)
        qrStep1Local.input.set(qr.data, deserialized_homogen_table)

        # Compute QR decomposition in step 1
        pres = qrStep1Local.compute()
        dataFromStep1ForStep2 = pres.get(qr.outputOfStep1ForStep2)
        serialized_1for2 = serializeNumericTable(dataFromStep1ForStep2)
        dataFromStep1ForStep3 = pres.get(qr.outputOfStep1ForStep3)
        serialized_1for3 = serializeNumericTable(dataFromStep1ForStep3)

        return (key, (serialized_1for2, serialized_1for3))

    dataFromStep1_RDD = dataRDD.map(mapper)

    def mapper_for_step3(tup):
        key, data_collections = tup
        dc1, dc2 = data_collections

        return (key, dc2)

    # Extract partial results for step 3
    dataFromStep1ForStep3_RDD = dataFromStep1_RDD.map(mapper_for_step3)

    def mapper_for_step2(tup):
        key, data_collections = tup
        dc1, dc2 = data_collections
        return (key, dc1)

    # Extract partial results for step 2
    dataFromStep1ForStep2_RDD = dataFromStep1_RDD.map(mapper_for_step2)

    return {'1for3': dataFromStep1ForStep3_RDD, '1for2':dataFromStep1ForStep2_RDD}


def computeStep2Master(dataFromStep1ForStep2_RDD):

    nBlocks = int(dataFromStep1ForStep2_RDD.count())

    dataFromStep1ForStep2_list = dataFromStep1ForStep2_RDD.collect()

    # Create an algorithm to compute QR decomposition on the master node
    qrStep2Master = qr.Distributed(step2Master, method=qr.defaultDense)

    for key, collection in dataFromStep1ForStep2_list:
        deserialized_collection = deserializeDataCollection(collection)
        qrStep2Master.input.add(qr.inputOfStep2FromStep1, key, deserialized_collection)

    # Compute QR decomposition in step 2
    pres = qrStep2Master.compute()

    inputForStep3FromStep2 = pres.getCollection(qr.outputOfStep2ForStep3)

    tup_list = []
    for key, collection in dataFromStep1ForStep2_list:
        dc = inputForStep3FromStep2[key]
        serialized_dc = serializeNumericTable(dc)
        tup_list.append((key, serialized_dc))

    # Make PairRDD from the list
    dataFromStep2ForStep3_RDD = sc.parallelize(tup_list, nBlocks)

    res = qrStep2Master.finalizeCompute()

    ntR = res.get(qr.matrixR)

    return {'ntR': ntR, 'from2_for3': dataFromStep2ForStep3_RDD}


def computeStep3Local(dataFromStep1ForStep3_RDD, dataFromStep2ForStep3_RDD):
    # Group partial results from steps 1 and 2
    dataForStep3_RDD = dataFromStep1ForStep3_RDD.cogroup(dataFromStep2ForStep3_RDD)

    def mapper(tup):
        key, collections = tup
        dc1, dc2 = collections

        ntQPi = dc1.__iter__().next()
        deserialized_ntQPi = deserializeDataCollection(ntQPi)
        ntPi = dc2.__iter__().next()
        deserialized_ntPi = deserializeDataCollection(ntPi)

        # Create an algorithm to compute QR decomposition on the master node
        qrStep3Local = qr.Distributed(step3Local, method=qr.defaultDense)
        qrStep3Local.input.set(qr.inputOfStep3FromStep1, deserialized_ntQPi)
        qrStep3Local.input.set(qr.inputOfStep3FromStep2, deserialized_ntPi)

        # Compute QR decomposition in step 3
        qrStep3Local.compute()
        result = qrStep3Local.finalizeCompute()

        Qi = result.get(qr.matrixQ)
        serialized_Qi = serializeNumericTable(Qi)

        return (key, serialized_Qi)

    return dataForStep3_RDD.map(mapper)


if __name__ == "__main__":

    # Create SparkContext that loads defaults from the system properties and the classpath and sets the name
    sc = SparkContext(conf=SparkConf().setAppName("Spark QR").setMaster('local[4]'))

    # Read from the distributed HDFS data set at a specified path
    dd = DistributedHDFSDataSet("/Spark/QR/data/")
    dataRDD = dd.getAsPairRDD(sc)

    # Compute QR decomposition for dataRDD
    result = runQR(dataRDD, sc)

    # Redirect stdout to a file for correctness verification
    stdout = sys.stdout
    sys.stdout = open('QR.out', 'w')

    # Print the results
    ntRPList = result['Q'].collect()
    for key, table in ntRPList:
        deserialized_table = deserializeNumericTable(table)
        printNumericTable(deserialized_table, "Q (2 first vectors from node #{}):".format(key), 2)

    printNumericTable(result['R'], "R:")

    # Restore stdout
    sys.stdout = stdout

    sc.stop()
